tests/benchmarks/LineInfoBenchmarks/FileInfoModels.cs

```diff
@@ -0,0 +1,218 @@
+        public ref LineInfo1Struct this[int index] { get { ArrayExtensions.EnsureSize(ref Lines, index); return ref Lines[index]; } }
+    }
+
+    public class FileInfo_Array_LineInfo1Class
+    {
+        public LineInfo1Class?[] Lines;
+        public FileInfo_Array_LineInfo1Class(int capactity) => Lines = new LineInfo1Class[capactity];
+        public LineInfo1Class? this[int index]
+        {
+            get { ArrayExtensions.EnsureSize(ref Lines, index); return Lines[index]; }
+            set { ArrayExtensions.EnsureSize(ref Lines, index); Lines[index] = value; }
+        }
+    }
+
+    public class FileInfo_Array_LineInfo2Struct
+    {
+        public LineInfo2Struct[] Lines;
+        public FileInfo_Array_LineInfo2Struct(int capactity) => Lines = new LineInfo2Struct[capactity];
+        public ref LineInfo2Struct this[int index] { get { ArrayExtensions.EnsureSize(ref Lines, index); return ref Lines[index]; } }
+    }
+
+    public class FileInfo_Array_LineInfo2Class
+    {
+        public LineInfo2Class?[] Lines;
+        public FileInfo_Array_LineInfo2Class(int capactity) => Lines = new LineInfo2Class[capactity];
+        public LineInfo2Class? this[int index]
+        {
+            get { ArrayExtensions.EnsureSize(ref Lines, index); return Lines[index]; }
+            set { ArrayExtensions.EnsureSize(ref Lines, index); Lines[index] = value; }
+        }
+    }
+
+    public class FileInfo_Array_HitsFlagsInt_Dictionary_BranchInfo1Struct
+    {
+        public const int IsLine = 1 << ((sizeof(int) * 8) - 1);
+        public const int IsBranch = 1 << ((sizeof(int) * 8) - 2);
+        public const int FlagMask = IsLine | IsBranch;
+        public const int HitsMask = ~FlagMask;
+        public int[] Lines;
+        public Dictionary<int, BranchInfo1Struct> Branches;
+        public FileInfo_Array_HitsFlagsInt_Dictionary_BranchInfo1Struct(int linesCapactity, int branchesCapacity)
+        {
+            Lines = new int[linesCapactity];
+            Branches = new(branchesCapacity);
+        }
+        public void CreateLine(LineInfo1Class other)
+        {
+            int index = other.LineNumber - 1;
+            int hits = other.Hits | IsLine;
+            if (other.TotalBranches != 0)
+            {
+                hits |= IsBranch;
+                Branches.Add(index + 1, new(other));
+            }
+            ArrayExtensions.EnsureSize(ref Lines, index);
+            Lines[index] = hits;
+        }
+        public void MergeLine(int index, int otherHits, BranchInfo1Struct otherBranchInfo)
+        {
+            if ((otherHits & IsLine) == 0) return;
+            ArrayExtensions.EnsureSize(ref Lines, index);
+            int hits = Lines[index] | IsLine;
+            BranchInfo1Struct branchInfo;
+            if ((hits & IsBranch) != 0)
+            {
+                if ((otherHits & IsBranch) != 0)
+                {
+                    branchInfo = Branches[index + 1];
```

Comment:
We could derive from Dictionary, implement AddOrUpdate, and use the internal FindValue which returns a ref TValue. Avoids the second lookup when reassigning. ConcurrentDictionary already has this.

Reply:
Sadly, that's internal and follow-ups like IsNullRef are internals as well and require unsafe context.

Reply:
Using ConcurrentDictionary however isn't an option, not just because of concurrency overhead, but the actual AddOrUpdate() function doesn't avoid the double lookup, it just is optimized for concurrency.